engine/engine.go

```diff
@@ -264,10 +274,13 @@ func (r *ruleEngine) RunRulesScoped(ctx context.Context, ruleSets []RuleSet, sco
 	}()
 
 	for _, rule := range otherRules {
+		newContext := ruleContext.Copy()
```

Comment:
This was needed because otherwise, scopes were added in parallel, and this caused race conditions.